cloud_Azure/terraform/module/examples/multiple_accounts_multiple_resource_group/profiles.py

```diff
@@ -0,0 +1,104 @@
+
+log = logging.getLogger(__name__)
+
+# pylint: disable=too-many-instance-attributes
+@dataclass
+class AzureProfile:
+    name: str
+    subscription_id: str
+    tenant_id: str
+    principal_id: str
+    principal_secret: str
+    location: str
+    resource_group_names: List[str]
+    storage_account_names: List[str]
+
+
+def load_profiles(file_path: str) -> List[AzureProfile]:
+    """
+    Return profile list on success
+    Return empty list if file_path doesn't exist or file contains no profiles
+    Raise RuntimeError on reading error/parsing error
+    """
+
+    config = configparser.ConfigParser()
+
+    try:
+        # note: config.read silently ignores the fact that file_path doesn't exist, resulting in empty config dict and no error reported
+        config.read(file_path)
+    except configparser.Error as err:
+        raise RuntimeError() from err
+
+    output_profiles: List[AzureProfile] = []
+    available_profiles: List[str] = [p for p in config.keys() if p != "DEFAULT"]  # skip ini file "DEFAULT" section
+    for profile_name in available_profiles:
+        profile = config[profile_name]
+        try:
+            resource_groups, storage_accounts = get_resource_groups_and_storage_accounts(profile_name, dict(profile))
+            azure_profile = AzureProfile(
+                name=profile_name,
+                subscription_id=profile["subscription_id"],
+                tenant_id=profile["tenant_id"],
+                location=profile["location"],
+                resource_group_names=resource_groups,
+                storage_account_names=storage_accounts,
+                principal_id=profile["principal_id"],
+                principal_secret=profile["principal_secret"],
+            )
+        except KeyError as err:
+            raise RuntimeError(f"Failed to read profile '{profile_name}' from file '{file_path}'") from err
+        output_profiles.append(azure_profile)
+
+    log.info("Loaded %d profile(s) from '%s'", len(output_profiles), file_path)
+    return output_profiles
+
+
+def get_resource_groups_and_storage_accounts(name: str, profile: Dict[str, Any]) -> Tuple[List[str], List[str]]:
+    def split_names(names: str) -> List[str]:
+        return [name.strip() for name in names.split(",")] if names else []
+
+    # get resource groups and storage accounts
+    resource_groups = split_names(profile["resource_group_names"])  # required
+    storage_accounts = split_names(profile.get("storage_account_names", ""))  # optional
+
+    if storage_accounts == []:
+        return (resource_groups, [])
+
+    # if storage_accounts are specified, there should be a Storage Account name specified for every Resource Group name
+    if len(storage_accounts) != len(resource_groups):
+        print_log(
+            f"In profile '{name}', 'storage_account_names' is specified but has different item count than 'resource_group_names'. Ignoring",
+            file=sys.stderr,
+            level=logging.WARNING,
```

Comment:
Custom names for storage accounts are optional.
Do you want to fail here?

Reply:
Based on my reading of the code, we rely on the order of entries in those 2 lists to associate storage account name with resource group. So, if the `storage_account_names` list has different number of entries than the`resource_group_names` list we cannot reliably associate custom storage account name with a resource group. So, unless I'm missing something, we need to fail here. I don't think it is necessary to support the use case where user provides custom storage account names only for some resource groups. I expect users to either let us auto-generate all storage account names or provide the all.

Reply:
Instead of failing we can also ignore invalid custom names and continue with generated ones, just giving the user a notification. Names can be corrected and updated later on.
So, should we:

1. Just ignore invalid names (meaning names will be generated for this profile)
2. Skip this specific profile and continue with next one
3. Stop the application telling user they need to correct custom names in profile such and such

?

Reply:
I prefer to stop the application. Ideally we should verify all names before making any changes, so that we do not leave the system in half-done state.

Reply:
Moved validation to separate function, outside file loading. Validation failure will stop azure_onboarder from executing any action. Will be reported in profiles tool "validate" mode. Will be ignored in "add" and "complete" mode - not all profiles need to be valid during the "profiles construction" phase